hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/pipeline/RatisPipelineUtils.java

```diff
@@ -117,4 +120,48 @@ static List<Pipeline> checkPipelineContainSameDatanodes(
                 p.sameDatanodes(pipeline)))
         .collect(Collectors.toList());
   }
+
+  /**
+   * Returns a map containing pipeline information which includes
+   * pipeline-id & the corresponding roles of all datanodes along with
+   * their respective datanode-id
+   * part of the pipeline.
+   *
+   * @param pipelines input pipeline
+   * @return map containing pipeline details
+   */
+  public static Map<String, Map<String, String[]>> pipelineLeaderFormat(
```

Comment:
We could simplify it and add the `hostname` or `role` or `uuid` as keys to improve user readability.

```suggestion
  public static Map<String, Map<String, String>> pipelineLeaderFormat(
```